collagraph/renderers/pyside_renderer.py

```diff
@@ -0,0 +1,181 @@
+from PySide6.QtWidgets import (
+    QBoxLayout,
+    QDialogButtonBox,
+    QFormLayout,
+    QGridLayout,
+    QMainWindow,
+    QSpacerItem,
+    QTabWidget,
+    QWidget,
+)
+
+from . import Renderer
+from .pyside import camel_case, dialog, name_to_type, tab, widget, window
+
+
+logger = logging.getLogger(__name__)
+
+
+def sorted_on_class_hierarchy(value):
+    # __mro__ is a tuple of all the classes that an class inherits. The longer the
+    # __mro__, the 'deeper' the subclass is, so we can use that to sort the classes
+    # to make the deepest class come first.
+    return dict(sorted(value.items(), key=lambda x: -len(x[0].__mro__)))
+
+
+INSERT_MAPPING = sorted_on_class_hierarchy(
+    {
+        QWidget: widget.insert,
+        QMainWindow: window.insert,
+        QDialogButtonBox: dialog.insert,
+        QTabWidget: tab.insert,
+    }
+)
+REMOVE_MAPPING = sorted_on_class_hierarchy(
+    {
+        QWidget: widget.remove,
+        QTabWidget: tab.remove,
+    }
+)
+SET_ATTR_MAPPING = sorted_on_class_hierarchy(
+    {
+        QWidget: widget.set_attribute,
+    }
+)
+
+# Cache for wrapped types
+WRAPPED_TYPES = {}
+
+LAYOUT = {
+    "Box": QBoxLayout,
+    "Grid": QGridLayout,
+    "Form": QFormLayout,
+}
+
+
+def not_implemented(self, *args, **kwargs):
+    raise NotImplementedError(type(self).__name__)
+
+
+class PySideRenderer(Renderer):
+    """PySide6 renderer."""
+
+    def create_element(self, type_name: str) -> Any:
+        """Create an element for the given type."""
+        # Make sure that an app exists before any widgets
+        # are created. Otherwise we might experience a
+        # hard segfault.
+        QtCore.QCoreApplication.instance() or QtWidgets.QApplication()
+
+        # TODO: how to do spacing / stretch?
+        if type_name in ["Spacing", "Stretch"]:
+            return QSpacerItem(0, 0)
+
+        # Create dynamic subclasses which implement `insert`, `set_attribute`
+        # and `remove` methods.
+        # The generated types are cached in WRAPPED_TYPES so they only have
+        # to be generated once and can be used in equality comparisons
+        if type_name in WRAPPED_TYPES:
+            return WRAPPED_TYPES[type_name]()
+
+        original_type = name_to_type(type_name)
+
+        attrs = {
+            "insert": not_implemented,
+            "remove": not_implemented,
+            "set_attribute": not_implemented,
+        }
+        for insert_class in INSERT_MAPPING:
+            if issubclass(original_type, insert_class):
+                attrs["insert"] = INSERT_MAPPING[insert_class]
+                break
+
+        for remove_class in REMOVE_MAPPING:
+            if issubclass(original_type, remove_class):
+                attrs["remove"] = REMOVE_MAPPING[remove_class]
+                break
+
+        for set_class in SET_ATTR_MAPPING:
+            if issubclass(original_type, set_class):
+                attrs["set_attribute"] = SET_ATTR_MAPPING[set_class]
+                break
```

Comment:
You can avoid a lot of the duplication here:

```suggestion
        attrs = {}
        maps = {
            "insert": INSERT_MAPPING,
            "remove": REMOVE_MAPPING,
            "set_attribute": SET_ATTR_MAPPING,
        }
        for key, map in maps.items():
            for cls, method in map.items():
                if issubclass(original_type, cls):
                    attrs[key] = method
                    break
            else:
                attrs[key] = not_implemented
```

Reply:
I had deliberately left this alone to keep it readable, but yes: this would do the trick! Thanks!